build/parseSpec.c

```diff
@@ -171,20 +171,64 @@ static int restoreFirstChar(rpmSpec spec)
     return 0;
 }
 
+static const struct lineTypes {
+    int id;
+    size_t textLen;
+    const char *text;
+    int withArgs;
+} specLineTypes[] = {
+    { LINE_IF,          LEN_AND_STR("%if")     , 1},
+    { LINE_IFARCH,      LEN_AND_STR("%ifarch") , 1},
+    { LINE_IFNARCH,     LEN_AND_STR("%ifnarch"), 1},
+    { LINE_IFOS,        LEN_AND_STR("%ifos")   , 1},
+    { LINE_IFNOS,       LEN_AND_STR("%ifnos")  , 1},
+    { LINE_ELSE,        LEN_AND_STR("%else")   , 0},
+    { LINE_ENDIF,       LEN_AND_STR("%endif")  , 0},
+    { LINE_INCLUDE,     LEN_AND_STR("%include"), 1},
+    { LINE_OTHER,       0, NULL                }
+};
+static const size_t numLineTypes = sizeof(specLineTypes)/sizeof(*specLineTypes);
+
+static int parseLineType(char *line)
+{
+    for (int i=0; i<(numLineTypes-1); i++) {
+	if (specLineTypes[i].withArgs) {
+	    if ( ISMACROWITHARG(line, specLineTypes[i].text, specLineTypes[i].textLen) )
+		return i;
+	} else {
+	    if ( ISMACRO(line, specLineTypes[i].text, specLineTypes[i].textLen) )
```

Comment:
Cosmetics: there's an extra space in around those ISMACRO*() "calls", and missing space in the for-loop number comparison.

The name specLineType seems overly generic here, when these are branch-related directives all. I'd suggest walking over the for-loop with a pointer and returning that if found, that way you don't need to separately index specLineTypes a hundred times in both this and the next commit. If you return NULL on not-found, you don't actually need LINE_OTHER type at all.